File: 27字符串的字典序打印/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
public class Solution {
    private ArrayList<String> ret = new ArrayList<>();
    public ArrayList<String> Permutation(String str) {
         if (str.length() == 0)
        return ret;
    char[] chars = str.toCharArray();
    Arrays.sort(chars);
    backtracking(chars, new boolean[chars.length], new StringBuilder());
    return ret;
}

private void backtracking(char[] chars, boolean[] hasUsed, StringBuilder s) {
    if (s.length() == chars.length) {
        ret.add(s.toString());
        return;
    }
    for (int i = 0; i < chars.length; i++) {
        if (hasUsed[i])
            continue;
        if (i != 0 && chars[i] == chars[i - 1] && !hasUsed[i - 1]) /* 保证不重复 */
            continue;
        hasUsed[i] = true;
        s.append(chars[i]);
        backtracking(chars, hasUsed, s);
        s.deleteCharAt(s.length() - 1);
        hasUsed[i] = false;
    }
}
}
/* java解法采取了完全不同的 一种思路，dfs，dfs的重点在于终止条件，避免重复
以及回溯，所以需要设置hasused变量来进行判断，以避免重复，同时遇到相等的字符
也要跳过，以避免重复，在一次的dfs结束以后要对hasused变量归零并删去最后一个
字符，以完成回溯 */

File: 27字符串的字典序打印/Solution.cpp
class Solution {

    
public:
    vector<string> Permutation(string str) {
        if(str.size()==0){
            return result;
        }
        PermutationC(str,0);
        sort(result.begin(),result.end());
        return result;
    }
private :
    vector<string> result ;
    void PermutationC(string str,int begin){
    //递归 结束的条件
    if (begin==str.size()){
        result.push_back(str);
        return;
    }
    for(int i=begin;i<str.size();i++){
        if (i!=begin&&str[begin]==str[i]) continue;
        swap(str[begin],str[i]);
        PermutationC(str,begin+1);
    }
}
};
/* 交换递归思路，现将第一个字符和后面所有的字符分别交换 得到若干种情况
然后再将交换后的数组进入递归，使第二个字符和后面的每一个字符交换
再递归，递归结束的条件是交换到了最后，则得到了一种结果，最后将
所有结果存入result中排序即可 */